File: content/develop/use-cases/pub-sub/rust/pubsub_hub.rs
//! Redis-backed pub/sub hub helper.
//!
//! Wraps PUBLISH / SUBSCRIBE / PSUBSCRIBE plus the introspection commands
//! (PUBSUB CHANNELS, PUBSUB NUMSUB, PUBSUB NUMPAT) into a small, named API
//! that:
//!
//! * publishes JSON-encoded messages to a channel and counts how many
//!   subscribers Redis reported delivering to
//! * creates named in-process subscribers that own a `redis::aio::PubSub`
//!   connection and a background task pumping messages into a ring buffer
//! * tracks per-channel publish counters and per-subscriber received counts
//!   for the demo UI
//!
//! Pub/sub has at-most-once delivery: a message that arrives while a
//! subscriber is disconnected is gone. If you need persistence or replay,
//! use Redis Streams instead.

use futures_util::StreamExt;
use redis::aio::ConnectionManager;
use redis::{AsyncCommands, Client, RedisResult};
use serde::Serialize;
use serde_json::Value;
use std::collections::{HashMap, VecDeque};
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};
use tokio::sync::oneshot;
use tokio::task::JoinHandle;

/// Maximum number of messages retained per subscriber for UI inspection.
pub const DEFAULT_BUFFER_SIZE: usize = 50;

/// Errors that can be raised by hub operations.
#[derive(Debug)]
pub enum HubError {
    /// A subscription with the same name already exists.
    DuplicateName(String),
    /// Caller passed an empty list of channels or patterns.
    EmptyTargets,
    /// Underlying Redis error.
    Redis(redis::RedisError),
}

impl std::fmt::Display for HubError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            HubError::DuplicateName(name) => {
                write!(f, "subscription named '{}' already exists", name)
            }
            HubError::EmptyTargets => {
                write!(f, "subscription requires at least one channel or pattern")
            }
            HubError::Redis(err) => write!(f, "{}", err),
        }
    }
}

impl std::error::Error for HubError {}

impl From<redis::RedisError> for HubError {
    fn from(err: redis::RedisError) -> Self {
        HubError::Redis(err)
    }
}

/// Message shape delivered to every Subscription's ring buffer. Pattern
/// subscriptions carry the original pattern in `pattern`; exact-match
/// subscriptions leave it None.
#[derive(Debug, Clone, Serialize)]
pub struct ReceivedMessage {
    pub channel: String,
    pub pattern: Option<String>,
    pub payload: Value,
    pub received_at_ms: i64,
}

/// A named in-process subscriber bound to one or more channels or patterns.
///
/// Each `Subscription` owns its own `redis::aio::PubSub` connection and a
/// background tokio task that pumps incoming messages into the ring buffer.
/// Subscriptions are independent: closing one does not affect another, even
/// if they share channels.
pub struct Subscription {
    name: String,
    targets: Vec<String>,
    is_pattern: bool,
    buffer: Mutex<VecDeque<ReceivedMessage>>,
    received: AtomicU64,
    closed: AtomicBool,
    buffer_size: usize,
    // Stop signal for the background task; wrapped so close() can take it.
    stop_tx: Mutex<Option<oneshot::Sender<()>>>,
    // Join handle for the background dispatch task.
    task: Mutex<Option<JoinHandle<()>>>,
}

impl Subscription {
    #[allow(dead_code)]
    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn targets(&self) -> Vec<String> {
        self.targets.clone()
    }

    pub fn is_pattern(&self) -> bool {
        self.is_pattern
    }

    pub fn received_total(&self) -> u64 {
        self.received.load(Ordering::Relaxed)
    }

    /// Return up to `limit` of the most recently received messages, newest first.
    pub fn messages(&self, limit: Option<usize>) -> Vec<ReceivedMessage> {
        let buf = self.buffer.lock().expect("subscription buffer poisoned");
        match limit {
            Some(n) => buf.iter().take(n).cloned().collect(),
            None => buf.iter().cloned().collect(),
        }
    }

    pub fn reset_received(&self) {
        let mut buf = self.buffer.lock().expect("subscription buffer poisoned");
        buf.clear();
        self.received.store(0, Ordering::Relaxed);
    }

    /// Signal the background task to exit and await it. Safe to call more than once.
    pub async fn close(&self) {
        if self.closed.swap(true, Ordering::SeqCst) {
            return;
        }
        let stop_tx = self
            .stop_tx
            .lock()
            .expect("subscription stop lock poisoned")
            .take();
        if let Some(tx) = stop_tx {
            // Receiver may have already exited; drop is harmless either way.
            let _ = tx.send(());
        }
        let task = self
            .task
            .lock()
            .expect("subscription task lock poisoned")
            .take();
        if let Some(handle) = task {
            let _ = handle.await;
        }
    }

    pub fn is_alive(&self) -> bool {
        !self.closed.load(Ordering::Relaxed)
    }

    /// Snapshot the subscription's metadata for the demo /state endpoint.
    pub fn to_json(&self) -> Value {
        serde_json::json!({
            "name": self.name,
            "targets": self.targets,
            "is_pattern": self.is_pattern,
            "received_total": self.received_total(),
            "alive": self.is_alive(),
        })
    }
}

/// In-process publisher and registry of named subscriptions.
#[derive(Clone)]
pub struct PubSubHub {
    inner: Arc<HubInner>,
}

struct HubInner {
    client: Client,
    publisher: ConnectionManager,
    introspect: ConnectionManager,
    buffer_size: usize,
    subscriptions: Mutex<HashMap<String, Arc<Subscription>>>,
    published_total: AtomicU64,
    delivered_total: AtomicU64,
    channel_published: Mutex<HashMap<String, u64>>,
}

/// Snapshot of hub-level counters. Serialised with snake_case keys to match
/// the reference Python `/state` payload.
#[derive(Debug, Clone, Serialize)]
pub struct Stats {
    pub published_total: u64,
    pub delivered_total: u64,
    pub received_total: u64,
    pub active_subscriptions: u64,
    pub channel_published: HashMap<String, u64>,
    pub pattern_subscriptions: i64,
}

impl PubSubHub {
    /// Build a hub from a `redis::Client`. The hub creates two
    /// `ConnectionManager`s: one for ordinary commands (PUBLISH, PUBSUB
    /// introspection) and a separate one for the same purposes from
    /// other code paths — sharing across handlers is cheap and safe
    /// because `ConnectionManager` is `Clone`.
    pub async fn new(client: Client, buffer_size: usize) -> RedisResult<Self> {
        let publisher = ConnectionManager::new(client.clone()).await?;
        let introspect = publisher.clone();
        Ok(Self {
            inner: Arc::new(HubInner {
                client,
                publisher,
                introspect,
                buffer_size,
                subscriptions: Mutex::new(HashMap::new()),
                published_total: AtomicU64::new(0),
                delivered_total: AtomicU64::new(0),
                channel_published: Mutex::new(HashMap::new()),
            }),
        })
    }

    /// Publish a JSON-encoded message to `channel` and return Redis'
    /// delivered count (the integer that `PUBLISH` itself returns).
    pub async fn publish(&self, channel: &str, message: &Value) -> RedisResult<i64> {
        let payload = serde_json::to_string(message)
            .unwrap_or_else(|_| serde_json::to_string(&Value::Null).unwrap());
        let mut conn = self.inner.publisher.clone();
        let delivered: i64 = conn.publish(channel, payload).await?;
        self.inner.published_total.fetch_add(1, Ordering::Relaxed);
        self.inner
            .delivered_total
            .fetch_add(delivered as u64, Ordering::Relaxed);
        let mut chan = self
            .inner
            .channel_published
            .lock()
            .expect("channel_published lock poisoned");
        *chan.entry(channel.to_string()).or_insert(0) += 1;
        Ok(delivered)
    }

    /// Register a named exact-match subscription on one or more channels.
    pub async fn subscribe(
        &self,
        name: &str,
        channels: Vec<String>,
    ) -> Result<Arc<Subscription>, HubError> {
        self.register(name, channels, false).await
    }

    /// Register a named pattern subscription on one or more glob patterns.
    pub async fn psubscribe(
        &self,
        name: &str,
        patterns: Vec<String>,
    ) -> Result<Arc<Subscription>, HubError> {
        self.register(name, patterns, true).await
    }

    async fn register(
        &self,
        name: &str,
        targets: Vec<String>,
        is_pattern: bool,
    ) -> Result<Arc<Subscription>, HubError> {
        if targets.is_empty() {
            return Err(HubError::EmptyTargets);
        }
        {
            let subs = self
                .inner
                .subscriptions
                .lock()
                .expect("subscriptions lock poisoned");
            if subs.contains_key(name) {
                return Err(HubError::DuplicateName(name.to_string()));
            }
        }

        // Each subscription owns its own pub/sub connection. Sharing one
        // connection across subscribers would couple their lifetimes —
        // closing one would close the channel for the others.
        let mut pubsub = self.inner.client.get_async_pubsub().await?;
        if is_pattern {
            for pattern in &targets {
                pubsub.psubscribe(pattern).await?;
            }
        } else {
            for channel in &targets {
                pubsub.subscribe(channel).await?;
            }
        }

        let (stop_tx, stop_rx) = oneshot::channel();
        let sub = Arc::new(Subscription {
            name: name.to_string(),
            targets: targets.clone(),
            is_pattern,
            buffer: Mutex::new(VecDeque::with_capacity(self.inner.buffer_size)),
            received: AtomicU64::new(0),
            closed: AtomicBool::new(false),
            buffer_size: self.inner.buffer_size,
            stop_tx: Mutex::new(Some(stop_tx)),
            task: Mutex::new(None),
        });

        // The spawned task owns `pubsub`. The Stream returned by
        // `pubsub.on_message()` borrows from `pubsub`, so the task body
        // is built around that local borrow — moving the PubSub out
        // mid-task would invalidate the stream.
        let sub_for_task = sub.clone();
        let targets_for_task = targets.clone();
        let handle = tokio::spawn(async move {
            // Block keeps the `stream` borrow (which holds &mut pubsub)
            // confined so we can call unsubscribe afterwards.
            {
                let mut stop_rx = stop_rx;
                let mut stream = pubsub.on_message();
                loop {
                    tokio::select! {
                        biased;
                        _ = &mut stop_rx => break,
                        next = stream.next() => {
                            match next {
                                Some(msg) => dispatch(&sub_for_task, &msg),
                                None => break,
                            }
                        }
                    }
                }
            }
            // Best-effort clean-up so Redis releases the channel slot
            // before the connection actually drops. Ignore errors: the
            // connection may already be closing.
            for target in &targets_for_task {
                if is_pattern {
                    let _ = pubsub.punsubscribe(target).await;
                } else {
                    let _ = pubsub.unsubscribe(target).await;
                }
            }
            drop(pubsub);
        });

        sub.task
            .lock()
            .expect("subscription task lock poisoned")
            .replace(handle);

        let mut subs = self
            .inner
            .subscriptions
            .lock()
            .expect("subscriptions lock poisoned");
        subs.insert(name.to_string(), sub.clone());
        Ok(sub)
    }

    /// Close and remove the named subscription. Returns true if it existed.
    pub async fn unsubscribe(&self, name: &str) -> bool {
        let sub = {
            let mut subs = self
                .inner
                .subscriptions
                .lock()
                .expect("subscriptions lock poisoned");
            subs.remove(name)
        };
        match sub {
            Some(s) => {
                s.close().await;
                true
            }
            None => false,
        }
    }

    pub fn subscriptions(&self) -> Vec<Arc<Subscription>> {
        let subs = self
            .inner
            .subscriptions
            .lock()
            .expect("subscriptions lock poisoned");
        subs.values().cloned().collect()
    }

    #[allow(dead_code)]
    pub fn get_subscription(&self, name: &str) -> Option<Arc<Subscription>> {
        let subs = self
            .inner
            .subscriptions
            .lock()
            .expect("subscriptions lock poisoned");
        subs.get(name).cloned()
    }

    /// PUBSUB CHANNELS — list channels with at least one exact-match subscriber.
    pub async fn active_channels(&self, pattern: &str) -> RedisResult<Vec<String>> {
        let mut conn = self.inner.introspect.clone();
        let channels: Vec<String> = redis::cmd("PUBSUB")
            .arg("CHANNELS")
            .arg(pattern)
            .query_async(&mut conn)
            .await?;
        let mut sorted = channels;
        sorted.sort();
        Ok(sorted)
    }

    /// PUBSUB NUMSUB — subscriber count per channel (exact-match only).
    pub async fn channel_subscriber_counts(
        &self,
        channels: &[String],
    ) -> RedisResult<HashMap<String, i64>> {
        if channels.is_empty() {
            return Ok(HashMap::new());
        }
        let mut conn = self.inner.introspect.clone();
        let mut cmd = redis::cmd("PUBSUB");
        cmd.arg("NUMSUB");
        for ch in channels {
            cmd.arg(ch);
        }
        let pairs: Vec<(String, i64)> = cmd.query_async(&mut conn).await?;
        Ok(pairs.into_iter().collect())
    }

    /// PUBSUB NUMPAT — total active pattern subscriptions across all clients.
    pub async fn pattern_subscriber_count(&self) -> RedisResult<i64> {
        let mut conn = self.inner.introspect.clone();
        let n: i64 = redis::cmd("PUBSUB")
            .arg("NUMPAT")
            .query_async(&mut conn)
            .await?;
        Ok(n)
    }

    /// Combined publish and subscribe counters plus the current registry size.
    pub async fn stats(&self) -> Stats {
        let subs = self.subscriptions();
        let received_total: u64 = subs.iter().map(|s| s.received_total()).sum();
        let channel_published = self
            .inner
            .channel_published
            .lock()
            .expect("channel_published lock poisoned")
            .clone();
        let pattern_subscriptions = self.pattern_subscriber_count().await.unwrap_or(0);
        Stats {
            published_total: self.inner.published_total.load(Ordering::Relaxed),
            delivered_total: self.inner.delivered_total.load(Ordering::Relaxed),
            received_total,
            active_subscriptions: subs.len() as u64,
            channel_published,
            pattern_subscriptions,
        }
    }

    /// Reset publish/delivery counters and clear every subscription's buffer.
    pub fn reset_stats(&self) {
        self.inner.published_total.store(0, Ordering::Relaxed);
        self.inner.delivered_total.store(0, Ordering::Relaxed);
        self.inner
            .channel_published
            .lock()
            .expect("channel_published lock poisoned")
            .clear();
        for sub in self.subscriptions() {
            sub.reset_received();
        }
    }

    /// Close every active subscription. Safe to call more than once.
    pub async fn shutdown(&self) {
        let subs: Vec<Arc<Subscription>> = {
            let mut map = self
                .inner
                .subscriptions
                .lock()
                .expect("subscriptions lock poisoned");
            let drained: Vec<Arc<Subscription>> = map.values().cloned().collect();
            map.clear();
            drained
        };
        for sub in subs {
            sub.close().await;
        }
    }
}

// The redis::Msg delivered by `on_message` carries the channel name as a
// `&str` borrow into the underlying connection buffer; we copy out into an
// owned ReceivedMessage so the lifetime is no longer tied to the stream.
fn dispatch(sub: &Arc<Subscription>, msg: &redis::Msg) {
    let channel = msg.get_channel_name().to_string();
    let pattern = msg.get_pattern::<String>().ok();
    let raw_payload: String = msg.get_payload().unwrap_or_default();
    let payload: Value = serde_json::from_str(&raw_payload).unwrap_or(Value::String(raw_payload));
    let received_at_ms = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0);
    let message = ReceivedMessage {
        channel,
        pattern,
        payload,
        received_at_ms,
    };
    let mut buf = sub.buffer.lock().expect("subscription buffer poisoned");
    if buf.len() == sub.buffer_size {
        buf.pop_back();
    }
    buf.push_front(message);
    sub.received.fetch_add(1, Ordering::Relaxed);
}

File: content/develop/use-cases/pub-sub/ruby/pubsub_hub.rb
# frozen_string_literal: true

# Redis-backed pub/sub hub helper.
#
# Wraps PUBLISH / SUBSCRIBE / PSUBSCRIBE plus the introspection
# commands (PUBSUB CHANNELS, PUBSUB NUMSUB, PUBSUB NUMPAT) into a
# small, named API that:
#
# * publishes JSON-encoded messages to a channel and counts how many
#   subscribers Redis reported delivering to
# * creates named in-process subscribers, each with its own Redis
#   connection and a background thread that pumps incoming messages
#   into a callback
# * tracks per-channel publish counters and per-subscriber received
#   counts for the demo UI
#
# Pub/sub has at-most-once delivery: a message that arrives while a
# subscriber is disconnected is gone. If you need persistence or
# replay, use Redis Streams instead.
#
# Ruby specifics
# --------------
# In redis-rb 5.x, `Redis#subscribe` (and `psubscribe`) blocks the
# calling thread *and* monopolises the connection. The Subscription
# class spawns one Thread per subscriber, each with its own Redis
# client. To tear down a subscription cleanly, the helper calls
# `unsubscribe` (or `punsubscribe`) on the same Redis client from the
# main thread — redis-rb forwards the command down the open
# subscriber connection, the subscribe block returns, and the worker
# thread exits without a sentinel message or a forced kill.

require 'json'
require 'redis'
require 'thread'

module PubSubHub
  # Message shape delivered to every Subscription callback. Pattern
  # subscriptions carry the original pattern in `pattern`; exact-match
  # subscriptions leave it nil.
  class ReceivedMessage
    attr_reader :channel, :pattern, :payload, :received_at_ms

    def initialize(channel:, pattern:, payload:, received_at_ms:)
      @channel = channel
      @pattern = pattern
      @payload = payload
      @received_at_ms = received_at_ms
    end

    def to_h
      {
        'channel' => @channel,
        'pattern' => @pattern,
        'payload' => @payload,
        'received_at_ms' => @received_at_ms
      }
    end
  end

  # A named in-process subscriber bound to one or more channels or
  # patterns. Each Subscription owns its own Redis client (and
  # therefore its own connection) and a background thread that pumps
  # incoming messages into the registered callback.
  class Subscription
    attr_reader :name, :targets

    def initialize(name:, hub:, targets:, is_pattern:, on_message: nil,
                   buffer_size: 50, redis_options: {})
      raise ArgumentError, 'Subscription requires at least one channel or pattern' if targets.nil? || targets.empty?

      @name = name
      @hub = hub
      @targets = targets.dup.freeze
      @is_pattern = is_pattern
      @on_message = on_message
      @buffer_size = buffer_size
      @buffer = []
      @lock = Mutex.new
      @received = 0
      @closed = false

      # Each Subscription gets its own Redis client. Sharing one across
      # subscribers would couple their lifetimes — closing one would
      # close the subscriber connection for the others.
      @redis = Redis.new(redis_options)

      # ready_latch lets the caller wait until SUBSCRIBE/PSUBSCRIBE has
      # actually been acknowledged by the server before publishing,
      # so the very first message can't race the subscription handshake.
      @ready_latch = Queue.new
      @thread = Thread.new { run }
      # Block until the first subscribe acknowledgement arrives. The
      # worker pushes `true` on a successful SUBSCRIBE/PSUBSCRIBE ack
      # and `false` if it errored before getting an ack — surface the
      # failure here instead of registering a dead subscription.
      ack = @ready_latch.pop
      unless ack
        begin
          @redis.close
        rescue StandardError
          # ignore — we're already aborting
        end
        raise "subscribe acknowledgement did not arrive for '#{@name}'"
      end
    end

    def is_pattern?
      @is_pattern
    end

    # Alias kept so callers can read either spelling.
    def is_pattern
      @is_pattern
    end

    def received_total
      @lock.synchronize { @received }
    end

    def messages(limit = nil)
      copy = @lock.synchronize { @buffer.dup }
      limit ? copy.first(limit) : copy
    end

    def reset_received
      @lock.synchronize do
        @buffer.clear
        @received = 0
      end
    end

    # Tear the subscription down by asking the server to unsubscribe
    # us. redis-rb sends the command down the same connection the
    # subscribe loop is reading from, so the subscribe block returns
    # naturally and the worker thread exits. After the join, close the
    # Redis client.
    def close
      return if @closed

      @closed = true
      begin
        if @is_pattern
          @redis.punsubscribe
        else
          @redis.unsubscribe
        end
      rescue StandardError
        # If the connection is already gone (e.g. server bounced), the
        # worker thread is on its way out anyway.
      end

      @thread.join(2)
      # If the worker is still stuck (rare), drop the connection
      # under it so the blocking read returns and the thread exits.
      if @thread.alive?
        begin
          @redis.close
        rescue StandardError
          # ignore
        end
        @thread.join(1)
        @thread.kill if @thread.alive?
      end

      begin
        @redis.close
      rescue StandardError
        # ignore
      end
    end

    def alive?
      !@closed && @thread.alive?
    end

    def to_h
      {
        'name' => @name,
        'targets' => @targets.to_a,
        'is_pattern' => @is_pattern,
        'received_total' => received_total,
        'alive' => alive?
      }
    end

    private

    def run
      if @is_pattern
        @redis.psubscribe(*@targets) do |on|
          on.psubscribe { |_pattern, _count| @ready_latch.push(true) }
          on.pmessage do |pattern, channel, raw|
            dispatch(channel: channel, pattern: pattern, raw: raw)
          end
          on.punsubscribe { |_pattern, _count| }
        end
      else
        @redis.subscribe(*@targets) do |on|
          on.subscribe { |_channel, _count| @ready_latch.push(true) }
          on.message do |channel, raw|
            dispatch(channel: channel, pattern: nil, raw: raw)
          end
          on.unsubscribe { |_channel, _count| }
        end
      end
    rescue StandardError
      # Make sure the constructor doesn't block forever if subscribe
      # blew up before its first ack.
      @ready_latch.push(false) if @ready_latch.empty?
    ensure
      # In case `close` wasn't called, mark the subscription closed
      # so #alive? reports correctly.
      @closed = true
    end

    def dispatch(channel:, pattern:, raw:)
      payload =
        begin
          JSON.parse(raw)
        rescue JSON::ParserError, TypeError
          raw
        end
      message = ReceivedMessage.new(
        channel: channel || '',
        pattern: pattern,
        payload: payload,
        received_at_ms: (Time.now.to_f * 1000).to_i
      )
      @lock.synchronize do
        @buffer.unshift(message)
        @buffer.pop while @buffer.length > @buffer_size
        @received += 1
      end
      if @on_message
        begin
          @on_message.call(message)
        rescue StandardError
          # User-supplied callback; don't let it kill the dispatch
          # thread or block the next message.
        end
      end
    end
  end

  # Publish/subscribe helper with publisher counters and subscriber
  # registry.
  class RedisPubSubHub
    attr_reader :redis

    def initialize(redis_client: nil, buffer_size: 50, redis_options: nil)
      @redis_options = redis_options || { host: 'localhost', port: 6379 }
      # Plain (non-subscriber) client for PUBLISH and PUBSUB
      # introspection commands.
      @redis = redis_client || Redis.new(@redis_options)
      @buffer_size = buffer_size

      @subs_lock = Mutex.new
      @subscriptions = {}

      @stats_lock = Mutex.new
      @published_total = 0
      @delivered_total = 0
      @channel_published = Hash.new(0)
    end

    # Publish a message to a channel. The body is JSON-encoded so
    # callers can pass hashes, arrays, or scalars without converting
    # on every call. The return value is what Redis itself reports:
    # the number of clients that were subscribed (directly or via
    # pattern) at the moment the message was fanned out.
    def publish(channel, message)
      payload = JSON.generate(message)
      delivered = @redis.publish(channel, payload).to_i
      @stats_lock.synchronize do
        @published_total += 1
        @delivered_total += delivered
        @channel_published[channel] += 1
      end
      delivered
    end

    def subscribe(name:, channels:, on_message: nil)
      register(name: name, targets: channels, is_pattern: false, on_message: on_message)
    end

    def psubscribe(name:, patterns:, on_message: nil)
      register(name: name, targets: patterns, is_pattern: true, on_message: on_message)
    end

    def unsubscribe(name)
      sub = @subs_lock.synchronize { @subscriptions.delete(name) }
      return false unless sub

      sub.close
      true
    end

    def subscriptions
      @subs_lock.synchronize { @subscriptions.values.dup }
    end

    def get_subscription(name)
      @subs_lock.synchronize { @subscriptions[name] }
    end

    # PUBSUB CHANNELS — list server-side channels with at least one
    # exact-match subscriber. Pattern subscribers are counted
    # separately via #pattern_subscriber_count.
    def active_channels(pattern = '*')
      result = @redis.pubsub('channels', pattern)
      Array(result).map(&:to_s).sort
    end

    # PUBSUB NUMSUB — subscriber count per channel. Returns an alternating
    # [channel, count, channel, count, ...] array in redis-rb; flatten it
    # into a hash keyed by channel name.
    def channel_subscriber_counts(channels)
      return {} if channels.nil? || channels.empty?

      flat = @redis.pubsub('numsub', *channels)
      result = {}
      Array(flat).each_slice(2) do |ch, count|
        result[ch.to_s] = count.to_i
      end
      result
    end

    # PUBSUB NUMPAT — server-wide active pattern subscription count.
    def pattern_subscriber_count
      @redis.pubsub('numpat').to_i
    end

    def stats
      published_total, delivered_total, channel_published =
        @stats_lock.synchronize do
          [@published_total, @delivered_total, @channel_published.dup]
        end
      subs = subscriptions
      received_total = subs.sum(&:received_total)
      {
        'published_total' => published_total,
        'delivered_total' => delivered_total,
        'received_total' => received_total,
        'active_subscriptions' => subs.length,
        'channel_published' => channel_published,
        'pattern_subscriptions' => pattern_subscriber_count
      }
    end

    def reset_stats
      @stats_lock.synchronize do
        @published_total = 0
        @delivered_total = 0
        @channel_published.clear
      end
      subscriptions.each(&:reset_received)
    end

    # Close every active subscription. Safe to call more than once.
    def shutdown
      subs = @subs_lock.synchronize do
        all = @subscriptions.values
        @subscriptions.clear
        all
      end
      subs.each(&:close)
    end

    private

    def register(name:, targets:, is_pattern:, on_message:)
      @subs_lock.synchronize do
        raise ArgumentError, "subscription named '#{name}' already exists" if @subscriptions.key?(name)

        sub = Subscription.new(
          name: name,
          hub: self,
          targets: targets,
          is_pattern: is_pattern,
          on_message: on_message,
          buffer_size: @buffer_size,
          redis_options: @redis_options
        )
        @subscriptions[name] = sub
        sub
      end
    end
  end
end
